Permits overriding of NodeJS local build container

conorsch opened the pull request:
Needed these changes over in https://github.com/freedomofpress/securethenews/pull/123, since the Gulp-based static asset management used in STN requires more packages to be installed within the Node build container. Although there's a `django_stack_npm_global_pkgs` var for providing global NPM packages, e.g. `gulp-cli`, those packages do _not_ get installed in the NodeJS container, and so aren't useful for the STN config. 

The default config for the NodeJS container image is the same—it's just overrideable now. Minor changes to the NodeJS build logic include setting `detach=no`, so we don't have to wait for a `touch`ed file, and also so Ansible reports errors if the commands failed. Also, I set `keep_volumes=no`, but I believe that might be superfluous for this particular use case. Trying to ensure that builds are always clean.

conorsch committed:
Lets node container block until done

We don't need to touch a file and wait for its existence; we set
detach=no on the container task, and Ansible will wait for the
containerized command to finish. It'll also report errors, which the
wait-for-file task didn't.

conorsch committed:
Makes Node builder Dockerfile configurable

Over in Secure The News, node-sass is used by gulp to build assets.
Node-sass must build bindings, so we need extra build dependencies.
Rather than add them here, in a generalized role used for multiple
projects, we can add it only to the STN provisioning config, as long as
the Dockerfile used for the Node logic can be overridden.

conorsch committed:
Forces fresh builds for local Node build container

Uses force=yes to rebuild a minimal image even if a matching tag was
already found locally. This is because the role now supports overriding
the Dockerfile used for building the NodeJS environment, so we don't
want a previous incarnation of the node builder to be in effect in a
custom environment. Forcing rebuilds every time slows down deployments
slightly, but at the cost of making them far more reliable.

Also sets keep_volumes=no on the Node container. In this specific
context, I believe the option is unwarranted: only local dirs are
mounted into the countainer. Still, we don't ever want previous volumes
sticking around: each role run should create a fresh volume and copy
everything in from scratch. Forcing volume removal ensures the build is
fresh every time.

msheiny commented on tasks/npm.yml:
```diff
@@ -33,13 +34,12 @@
-    - name: Wait for node logic to finish up
-      wait_for:
-        path: "{{ tmp_dir }}/npm.done"
+        keep_volumes: no
```
what about destroying `node_modules` folder everytime for `ci`? Or having a flag to do that ? That would really ensure that everything is getting rebuilt fresh each time in a local environment

msheiny commented on tasks/npm.yml:
```diff
@@ -33,13 +34,12 @@
+        command: /bin/ash -c "{{ django_stack_npm_install_cmd }}"
         user: "{{ 'node' if (lookup('pipe','id -u') == '1000') else omit }}"
+        # Block until command is finished; report failure on error
+        detach: no
```
Noiiceee ❤️

msheiny commented on tasks/npm.yml:
```diff
@@ -19,9 +19,10 @@
       docker_image:
         name: "node_grsec:{{ django_stack_node_ver }}"
-        dockerfile: "{{ role_path }}/docker/NodeDockerfile"
+        dockerfile: "{{ django_stack_node_dockerfile }}"
```
We discussed this in person... regarding the trade-off of trying to do this via templating versus specifying a new dockerfile. Agreed that this approach is path of least resistance for now though and the right move to keep moving forward 👍

msheiny reviewed (approved):
Looking solid buddy! I didn't test locally but I'm relying CI ❤️  and visual scan through.

conorsch committed:
Removes grsecurity requirement for NodeJS override

Didn't notice on my local workstation because I'm running a
grsecurity-patched kernel. In CI, however, where there's no grsec, the
role failed loudly.
